In [3]:
import pandas as pd
import numpy as np

In [77]:
df=pd.read_csv("/content/Advanced_pandas_sales_dataset.csv")
df.head(10)

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
6,1007,2024-06-24,Dev,Chennai,East,Monitor,Electronics,1170,5,0.20,22.83
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
9,1010,2024-12-25,Pooja,Ahmedabad,North,Phone,Electronics,2711,8,0.00,477.32


In [4]:
'''
Q1 Dataset Exploration

Perform the following exploratory steps:

Display the first 10 rows

Display the dataset structure using info()

Find the number of rows and columns

Identify the data types of each column'''

# Display the first 10 rows
df.head(10)

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
6,1007,2024-06-24,Dev,Chennai,East,Monitor,Electronics,1170,5,0.20,22.83
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
9,1010,2024-12-25,Pooja,Ahmedabad,North,Phone,Electronics,2711,8,0.00,477.32


In [5]:
#Display the dataset structure using info()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230 entries, 0 to 229
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Order_ID    230 non-null    int64  
 1   Order_Date  230 non-null    object 
 2   Customer    230 non-null    object 
 3   City        230 non-null    object 
 4   Region      230 non-null    object 
 5   Product     230 non-null    object 
 6   Category    230 non-null    object 
 7   Sales       230 non-null    int64  
 8   Quantity    230 non-null    int64  
 9   Discount    230 non-null    float64
 10  Profit      230 non-null    float64
dtypes: float64(2), int64(3), object(6)
memory usage: 19.9+ KB


In [12]:
# Find the number of rows and columns

len(df)

# or we can use
df.shape

(230, 11)

In [13]:
# Identify the data types of each column

df.dtypes

,0
Order_ID,int64
Order_Date,object
Customer,object
City,object
Region,object
Product,object
Category,object
Sales,int64
Quantity,int64
Discount,float64


In [ ]:
'''
Q2 Sales Filtering Analysis

Filter the dataset to display:

Orders where Sales > 3500

Orders where Region = "East" and Sales > 2500

Explain what insights you observe.'''


In [14]:
df[(df.Sales>3500)]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
10,1011,2024-06-30,Neha,Delhi,South,Printer,Office,5977,6,0.00,1578.32
...,...,...,...,...,...,...,...,...,...,...,...
219,1220,2024-03-22,Meera,Bangalore,West,Phone,Electronics,5986,7,0.05,986.70
222,1223,2024-03-04,Meera,Ahmedabad,South,Phone,Electronics,4356,6,0.10,416.87
223,1224,2024-01-07,Dev,Bangalore,East,Tablet,Electronics,4115,4,0.05,870.03
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00


In [16]:
# Orders where Region = "East" and Sales > 2500


df[(df.Region=='East') &(df.Sales>2500)]


,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
13,1014,2024-03-18,Anita,Bangalore,East,Mouse,Accessories,3616,4,0.05,463.93
21,1022,2024-01-14,Nitin,Chennai,East,Monitor,Electronics,4827,6,0.05,912.34
27,1028,2024-07-21,Dev,Kolkata,East,Tablet,Electronics,4054,6,0.00,1013.83
33,1034,2024-12-15,Nitin,Ahmedabad,East,Tablet,Electronics,4378,2,0.20,-104.30
34,1035,2024-02-20,Nitin,Ahmedabad,East,Laptop,Electronics,2535,4,0.10,188.15
41,1042,2024-07-15,Rohan,Chennai,East,Laptop,Electronics,3268,8,0.15,295.15
43,1044,2024-09-20,Kavya,Hyderabad,East,Printer,Office,2618,7,0.05,454.64


In [ ]:
#df.sort_values(by='Sales',ascending=False).head(5).Customer

In [ ]:
'''
Q3 Customer Purchase Analysis

Identify the top 5 customers based on total sales.


Hint: Use GroupBy with aggregation.
'''



In [48]:
# Identify the top 5 customers based on total sales

#df.sort_values(by='Sales',ascending=False).head(5).Customer

df.groupby('Customer')['Sales'].sum().sort_values(ascending=False).head(5)
#  s4econd method
# df1=df['Sales'].sum().sort_values(ascending=false).head(5)


,Sales
Customer,
Nitin,59820
Raj,58416
Rahul,52291
Dev,48500
Isha,46818


In [ ]:
'''
Q4 Regional Sales Performance

Calculate the total sales generated by each region.

Then identify:

Which region generated the highest revenue?'''



In [58]:
#Calculate the total sales generated by each region.
df.groupby('Region')['Sales'].sum()

,Sales
Region,
East,232675
North,158210
South,183108
West,170227


In [59]:
#hich region generated the highest revenue?
df.groupby('Region')['Sales'].sum().sort_values(ascending=False).head(1)

,Sales
Region,
East,232675


so the region **East** is generating the highest sales among all the regions


In [ ]:
''' Q5 Product Performance Analysis

Calculate the total quantity sold for each product.

Which product appears to be the most popular?'''

In [61]:
# Calculate the total quantity sold for each product.
df.groupby('Product')['Quantity'].sum()

,Quantity
Product,
Headphones,115
Keyboard,110
Laptop,140
Monitor,156
Mouse,122
Phone,151
Printer,108
Tablet,150


In [62]:
# Which product appears to be the most popular
df.groupby('Product')['Quantity'].sum().sort_values(ascending=False).head(1)

,Quantity
Product,
Monitor,156


so the product Monitor is most popular with the highest quantity sold i.e 156

In [68]:
'''Q6 Multi-Level Grouping

Group the dataset by: Region and Category'''

df.groupby(['Region','Category'])['Sales'].sum()

#Calculate the total sales for each combination.
df.groupby(['Region','Category'])['Sales'].sum()

#What insights can you derive from this analysis?
df.groupby(['Region','Category'])['Sales'].sum()

Region  Category   
East    Accessories     57646
        Electronics    160907
        Office          14122
North   Accessories     62052
        Electronics     70023
        Office          26135
South   Accessories     80914
        Electronics     86690
        Office          15504
West    Accessories     64811
        Electronics     84128
        Office          21288
Name: Sales, dtype: int64

from the above result it can be easily interpreted thatregion east is performing good in comparison to other regions
also in terms of categories, electronics is the leading performer in all the regions followed by Accessories and Office **bold text**

In [97]:
'''
Q7 Pivot Table Creation

Create a pivot table showing:

Rows → Region

Columns → Product

Values → Total Sales

Which region sells the most laptops?
answer is east region

'''

pivot_df = df.pivot_table(
    index='Region',
    columns='Product',
    values='Sales',
    aggfunc='sum'
)



print(pivot_df)


Product  Headphones  Keyboard  Laptop  Monitor  Mouse  Phone  Printer  Tablet
Region                                                                       
East          23899     10845   55568    31299  22902  39667    14122   34373
North         15450     20538    4540    14922  26064  30535    26135   20026
South         24111     23480   23029    19167  33323  18218    15504   26276
West          17159     38499   10826    18178   9153  27850    21288   27274


In [86]:

# Q8 Profitability Analysis

# Calculate the average profit for each product category.
df.groupby('Category')['Profit'].mean()
# Which category is most profitable?
df.groupby('Category')['Profit'].mean().sort_values(ascending=False).head(1)

,Profit
Category,
Office,471.938462


auestion 8 Which category is most profitable?
answer:-
the office **category** is more profitable.


In [ ]:
'''Q9 Discount Impact Study

Analyze the relationship between discount and profit.'''


#Calculate the average profit for each discount level
df.groupby('Discount')['Profit'].mean()
# Identify whether higher discounts lead to lower profit

In [88]:
df.head(5)

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30


In [89]:
df.groupby('Discount')['Profit'].mean()

,Profit
Discount,
0.00,751.210588
0.05,531.368039
0.10,346.460513
0.15,237.191400
0.20,90.026923


In [92]:
#Identify whether higher discounts lead to lower profit
we can easily  see that as the discount is increasing the profit is decreasing significantly

AttributeError: 'Series' object has no attribute 'relationship'

In [96]:
# Q10 End-to-End Data Wrangling Task

#Filter the dataset where Sales > 3000
df[df.Sales>3000]

# Group data by Region and Product
df[df.Sales>3000].groupby(['Region','Product'])['Sales'].sum()

#Calculate total sales
df.Sales.sum()

#Identify the top 5 customers
df.groupby('Customer')['Sales'].sum().sort_values(ascending=False).head(5)

#Reshape the result into a pivot table
df.pivot_table(index='Region',columns='Product',values='Sales',aggfunc='sum')
#Explain the business insights obtained from this analysis

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,23899,10845,55568,31299,22902,39667,14122,34373
North,15450,20538,4540,14922,26064,30535,26135,20026
South,24111,23480,23029,19167,33323,18218,15504,26276
West,17159,38499,10826,18178,9153,27850,21288,27274
